include/blast/math/dense/StaticMatrix.hpp

```diff
@@ -0,0 +1,184 @@
+// Copyright (c) 2019-2020 Mikhail Katliar All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+#pragma once
+
+#include <blast/math/Forward.hpp>
+#include <blast/math/StorageOrder.hpp>
+#include <blast/math/Simd.hpp>
+#include <blast/math/TypeTraits.hpp>
+#include <blast/system/CacheLine.hpp>
+#include <blast/util/NextMultiple.hpp>
+#include <blast/util/Types.hpp>
+
+#include <initializer_list>
+
+
+namespace blast
+{
+    /// @brief Matrix with statically defined size.
+    ///
+    /// @tparam T element type of the matrix
+    /// @tparam M number of rows
+    /// @tparam N number of columns
+    /// @tparam SO storage order
+    template <typename T, size_t M, size_t N, bool SO>
+    class StaticMatrix
+    {
+    public:
+        using ElementType = T;
+        static bool constexpr storageOrder = SO;
+
+
+        StaticMatrix() noexcept
+        {
+            // Initialize padding elements to 0 to prevent denorms in calculations.
+            // Denorms can significantly impair performance, see https://github.com/giaf/blasfeo/issues/103
+            std::fill_n(v_, capacity_, T {});
+        }
+
+
+        StaticMatrix(T const& v) noexcept
+        {
+            std::fill_n(v_, capacity_, v);
+        }
+
+
+        constexpr StaticMatrix(std::initializer_list<std::initializer_list<T>> list)
+        {
+            std::fill_n(v_, capacity_, T {});
+
+            if (list.size() != M || determineColumns(list) > N)
+                throw std::invalid_argument {"Invalid setup of static matrix"};
```

Comment:
Does this work with both row and column major?

Reply:
Because for row major `list.size()` is `M` but for column major it is `N`

Reply:
It is how the matrix is construct from an initializer list. It is supposed to work in the same way regardless of the storage order, so I think it is correct.

Reply:
TODO: add function documentation